File: windows/WallpaperInfoApp/WallpaperInfoApp/WallpaperServiceHandler.cs
using System;
using System.Collections.Generic;
using System.Drawing;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace WallpaperInfoApp {

	class WallpaperServiceHandler {

		public WallpaperServiceHandler(WallpaperService wallpaperService) {
			_service = wallpaperService;
		}

		public static WallpaperServiceHandler incomingHandler(WallpaperServiceConnection incomingMessenger) {
			WallpaperServiceInfo wpsi = WallpaperServiceInfo.getInstance();
			wpsi.getWallpaperService()._wallpaperServiceHandler._incomingMessenger = incomingMessenger;
			return wpsi.getWallpaperService()._wallpaperServiceHandler;
		}
    
		public static void startService() {
			WallpaperServiceInfo.getInstance().getWallpaperService().startService();
		}
    
		public static void stopService() {
			WallpaperServiceInfo.getInstance().getWallpaperService().stopService();
		}
	
		public void replyToClient(Bitmap thumbnail, WPath currentWPath, bool pause) {
			if (_incomingMessenger != null) {
				_incomingMessenger._wallpaperInfoUI.broadcastReceiver(thumbnail, currentWPath, pause);
			}
		}
    
		public void handleMessage(int command, int intOption, object objectparm = null) {
		
			WallpaperServiceInfo wpsi = WallpaperServiceInfo.getInstance();
		
			switch (command) {
			
			case MSG.REQUEST_INFO:
				BPUtil.BPLog("WallpaperServiceHandler - MSG.REQUEST_INFO");
				break;

			case MSG.PAUSE:
				BPUtil.BPLog("WallpaperServiceHandler - MSG.PAUSE");
				wpsi.setPause(intOption == 1);
				_service.pause_resume_service();
				break;

			case MSG.PREVIOUS:
				BPUtil.BPLog("WallpaperServiceHandler - MSG.PREVIOUS");
				_service.restartTimer();
				_service.naviageWallpaper(offset: -1);
				break;

			case MSG.NEXT:
				BPUtil.BPLog("WallpaperServiceHandler - MSG.NEXT");
				_service.restartTimer();
				_service.naviageWallpaper(offset: 1);
				break;

			case MSG.SET_ROOT:
				BPUtil.BPLog("WallpaperServiceHandler - MSG.SET_ROOT");
				String path = (String)objectparm;
				_service.setNewRootPath(newRootPath: path);
				break;

			case MSG.ACTIVITY_REPORT:
				BPUtil.BPLog("WallpaperServiceHandler - MSG.ACTIVITY_REPORT");
				break;

			case MSG.SET_WALLPAPER:
				BPUtil.BPLog("WallpaperServiceHandler - MSG.SET_WALLPAPER");
				break;

			case MSG.SET_INTERVAL:
				BPUtil.BPLog("WallpaperServiceHandler - MSG.SET_INTERVAL");
				if (intOption > 0) {
					wpsi.setInterval(intOption);
					_service.setInterval(interval: intOption);
					_service.restartTimer();
				}
				break;

			case MSG.SERVICE_INFO:
				BPUtil.BPLog("WallpaperServiceHandler - MSG.SERVICE_INFO");
				break;

			case MSG.SET_THEME:
				BPUtil.BPLog("WallpaperServiceHandler - MSG.SET_THEME");
				_service.updateServiceTheme(ThemeMethods.rawValue(intOption));
				_service.restartTimer();
				break;

			case MSG.PREVIOUS_THEME:
				BPUtil.BPLog("WallpaperServiceHandler - MSG.PREVIOUS_THEME");
				_service.updateServiceThemeWithPrevious();
				break;

			case MSG.NEXT_THEME:
				BPUtil.BPLog("WallpaperServiceHandler - MSG.NEXT_THEME");
				_service.updateServiceThemeWithNext();
				break;

			case MSG.CUSTOM_CONFIG:
				BPUtil.BPLog("WallpaperServiceHandler - MSG.CUSTOM_CONFIG");
				String customConfigString = (String)objectparm;
				_service.updateServiceCustomConfig(customConfigString);
				_service.restartTimer();
				break;

			default:
				BPUtil.BPLog("WallpaperServiceHandler - Invalid Command Error");
				break;
			}
		}
	
		public void unbind() {
			_incomingMessenger = null;
		}

		private WallpaperServiceConnection _incomingMessenger = null;
		private WallpaperService _service = null;
	}
}

File: apps/desktop-ubuntu/WallpaperInfoApp/WallpaperServiceHandler.py
#!/usr/bin/env python3

import AppUtil as a
import WallpaperServiceInfo as w
import ThemeInfo as t
import BPUtil

class WallpaperServiceHandler:

    def __init__(self, wallpaperService):
        self._service = wallpaperService
        self._incomingMessenger = None

    def replyToClient(self, thumbnail, currentWPath, pause):
        if self._incomingMessenger != None:
            self._incomingMessenger._wallpaperInfoUI.broadcastReceiver(thumbnail, currentWPath, pause)

    def handleMessage(self, msg, intOption = 0, objectOption = None):
        wpsi = w.WallpaperServiceInfo();
        if msg == a.MSG.CUSTOM_CONFIG:
            BPUtil.BPLog("WallpaperServiceHandler - MSG.CUSTOM_CONFIG")
            customConfigString = objectOption;
            self._service.updateServiceCustomConfig(customConfigString)
            self._service.restartTimer()
        elif msg == a.MSG.REQUEST_INFO:
            BPUtil.BPLog("WallpaperServiceHandler - MSG.REQUEST_INFO")
        elif msg == a.MSG.PAUSE:
            BPUtil.BPLog("WallpaperServiceHandler - MSG.PAUSE")
            wpsi.setPause(intOption == 1)
            self._service.pause_resume_service()
        elif msg == a.MSG.PREVIOUS:
            BPUtil.BPLog("WallpaperServiceHandler - MSG.PREVIOUS")
            self._service.restartTimer()
            self._service.naviageWallpaper(-1)
        elif msg == a.MSG.NEXT:
            BPUtil.BPLog("WallpaperServiceHandler - MSG.NEXT")
            self._service.restartTimer()
            self._service.naviageWallpaper(1)
        elif msg == a.MSG.SET_INTERVAL:
            BPUtil.BPLog("WallpaperServiceHandler - MSG.SET_INTERVAL")
            if intOption > 0:
                wpsi.setInterval(intOption)
                self._service.setInterval(intOption)
                self._service.restartTimer()
        elif msg == a.MSG.SET_THEME:
            BPUtil.BPLog("WallpaperServiceHandler - MSG.SET_THEME")
            self._service.updateServiceTheme(intOption)
            self._service.restartTimer()
        elif msg == a.MSG.PREVIOUS_THEME:
            BPUtil.BPLog("WallpaperServiceHandler - MSG.PREVIOUS_THEME")
            self._service.updateServiceThemeWithPrevious();
        elif msg == a.MSG.NEXT_THEME:
            BPUtil.BPLog("WallpaperServiceHandler - MSG.NEXT_THEME")
            self._service.updateServiceThemeWithNext();
        elif msg == a.MSG.SET_MODE:
            BPUtil.BPLog("WallpaperServiceHandler - MSG.SET_MODE")
        elif msg == a.MSG.SET_SAVER:
            BPUtil.BPLog("WallpaperServiceHandler - MSG.SET_SAVER")
        elif msg == a.MSG.TOGGLE_PAUSE:
            BPUtil.BPLog("WallpaperServiceHandler - MSG.TOGGLE_PAUSE")
        elif msg == a.MSG.SET_ROOT:
            BPUtil.BPLog("WallpaperServiceHandler - MSG.SET_ROOT")
            self._service.setNewRootPath(objectOption)
        elif msg == a.MSG.SET_WALLPAPER:
            BPUtil.BPLog("WallpaperServiceHandler - MSG.SET_WALLPAPER")
        else:
            BPUtil.BPLog("WallpaperServiceHandler - Invalid MSG Error")

    def unbind(self):
        self._incomingMessenger = None


def incomingHandler(incomingMessenger):
    wpsi = w.WallpaperServiceInfo();
    wpsi.getWallpaperService()._wallpaperServiceHandler._incomingMessenger = incomingMessenger;
    return wpsi.getWallpaperService()._wallpaperServiceHandler

def startService():
    w.WallpaperServiceInfo().getWallpaperService().startService()

def stopService():
    w.WallpaperServiceInfo().getWallpaperService().stopService()

#w_handler = WallpaperServiceHandler(w.WallpaperServiceInfo().getWallpaperService())
#w_handler.handleMessage(a.MSG.REQUEST_INFO, 2)
